cmd/agent/api/security.go

```diff
@@ -0,0 +1,114 @@
+package api
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"net/http"
+
+	"github.com/DataDog/datadog-agent/pkg/api/security"
+	"github.com/DataDog/datadog-agent/pkg/api/util"
+	grpc_auth "github.com/grpc-ecosystem/go-grpc-middleware/auth"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type contextKey int
+
+const (
+	contextKeyTokenInfoID contextKey = iota
+)
+
+var (
+	tlsKeyPair  *tls.Certificate
+	tlsCertPool *x509.CertPool
+	tlsAddr     string
+)
+
+// validateToken - validates token for legacy API
+func validateToken(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if err := util.Validate(w, r); err != nil {
+			return
+		}
+		next.ServeHTTP(w, r)
+	})
+}
+
+// parseToken parses the token and validate it for our gRPC API, it returns an empty
+// struct and an error or nil
+func parseToken(token string) (struct{}, error) {
+	if token != util.GetAuthToken() {
+		return struct{}{}, errors.New("Invalid session token")
+	}
+
+	// Currently this empty struct doesn't add any information
+	// to the context, but we could potentially add some custom
+	// type.
+	return struct{}{}, nil
+}
+
+//grpcAuthFunc is a middleware (interceptor) that extracts and verifies token from header
+func grpcAuth(ctx context.Context) (context.Context, error) {
+
+	token, err := grpc_auth.AuthFromMD(ctx, "Bearer")
+	if err != nil {
+		return nil, err
+	}
+
+	tokenInfo, err := parseToken(token)
+	if err != nil {
+		return nil, status.Errorf(codes.Unauthenticated, "invalid auth token: %v", err)
+	}
+
+	// do we need this at all?
+	newCtx := context.WithValue(ctx, contextKeyTokenInfoID, tokenInfo)
```

Comment:
Because it's kinda sensitive, I would not share this in the context until it is effectively needed by a next interceptor. WDYT?

Reply:
Nevermind I got the idea. I've seen later that this is just an empty context struct for now (misread and thought this was the bearer token...)